src/_pytest/terminal.py

```diff
@@ -803,31 +806,36 @@ def getreports(self, name):
         return values
 
     def summary_warnings(self):
-        if self.hasopt("w"):
-            all_warnings = self.stats.get(
-                "warnings"
-            )  # type: Optional[List[WarningReport]]
-            if not all_warnings:
-                return
+        if not self.hasopt("w"):
+            return
 
-            final = hasattr(self, "_already_displayed_warnings")
-            if final:
-                warning_reports = all_warnings[self._already_displayed_warnings :]
-            else:
-                warning_reports = all_warnings
-            self._already_displayed_warnings = len(warning_reports)
-            if not warning_reports:
-                return
+        all_warnings = self.stats.get("warnings")  # type: Optional[List[WarningReport]]
+        if not all_warnings:
+            return
 
-            reports_grouped_by_message = (
-                collections.OrderedDict()
-            )  # type: collections.OrderedDict[str, List[WarningReport]]
-            for wr in warning_reports:
-                reports_grouped_by_message.setdefault(wr.message, []).append(wr)
+        final = hasattr(self, "_already_displayed_warnings")
```

Comment:
I understand the code was already like this, but how about we initialize this with `0` during `__init__`? besides using dynamic attributes kind of a smell which has brought bugs in the past, this will also simplify the code because you won't need the `if` anymore:

```python
warning_reports = all_warnings[self._already_displayed_warnings :]
```

Reply:
@nicoddemus 
Feel free to suggest an edit.